onnx/defs/training/defs.cc

```diff
@@ -324,4 +324,134 @@ ONNX_TRAINING_OPERATOR_SET_SCHEMA(
+            "gradient of \"X_2\", accumulated squared gradient of \"X_1\", accumulated squared gradient of \"X_2\"].",
+            "T3",
+            OpSchema::Variadic,
+            false)
+        .Output(
+            0,
+            "outputs",
+            "It sequentially contains the new values of optimized tensors and then the new "
+            "values of accumulated gradient. For example, if two tensor \"X_1\" and \"X_2\" are "
+            "optimized, the output list would be [new value of \"X_1,\" new value of \"X_2\" "
+            "new accumulated squared gradient of \"X_1\", new accumulated squared gradient of \"X_2\"].",
+            "T2",
+            OpSchema::Variadic,
+            false)
+        .Attr(
+            "epsilon",
+            "Small scalar to avoid dividing by zero.",
+            AttributeProto::FLOAT,
+            1e-6f)
+        .Attr(
+            "decay_factor",
+            "The decay factor of learning rate after one update."
+            "The effective learning rate is computed by r = R / (1 + T * decay_factor). "
+            "Default to 0 so that increasing update counts doesn't reduce the learning rate.",
+            AttributeProto::FLOAT,
+            0.0f)
+        .Attr(
+            "norm_coefficient",
+            "Regularization coefficient in 0.5 * norm_coefficient * ||X||_2^2. Default to 0, "
+            "which means no regularization.",
+            AttributeProto::FLOAT,
+            0.0f)
+        .TypeConstraint(
+            "T1",
+            {"tensor(float)", "tensor(double)"},
+            "Constrain input types to float scalars.")
+        .TypeConstraint(
+            "T2",
+            {"tensor(int64)"},
+            "Constrain output types to 64-bit integer scalars.")
+        .TypeConstraint(
+            "T3",
+            {"tensor(float)", "tensor(double)"},
+            "Constrain input types to float tensors.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+            // In comments below, we assume that the input list is
+            // [R, T, X1, X2, G1, G2, H1, H2] and the output list is
+            // [X1_new, X2_new, H1_new, H2_new].
+
+            // Compute the number of tuples (X, G, H).
+            auto num_optimized_tensors = (ctx.getNumInputs() - 2) / 3;
+            for (size_t i = 0; i < num_optimized_tensors; ++i) {
+              // Pass X1's and X2's shapes to X1_new and X2_new, respectively.
+              size_t i_in = 2 + i;
+              size_t i_out = i;
+              propagateElemTypeFromInputToOutput(ctx, i_in, i_out);
+              propagateShapeFromInputToOutput(ctx, i_in, i_out);
+
+              // Pass H1's and H2's shapes to H1_new and H2_new, respectively.
+              i_in = 2 + 2 * num_optimized_tensors + i;
+              i_out = i + num_optimized_tensors;
+              propagateElemTypeFromInputToOutput(ctx, i_in, i_out);
+              propagateShapeFromInputToOutput(ctx, i_in, i_out);
+            }}));
```

Comment:
It seems that this operator could be decomposed into more primitive operations. Can we add a `FunctionBody` for it as well?

Reply:
Yes... but we need to think about how to process variable-length inputs using a static graph.